### ISSUE TITLE
Performance: Improve performance of FastBuffer/walker

### ISSUE THREAD

**Issue by extr0py** [extr0py/oni#229]
Looking at the latest profile of typing in ONI after the plugin fix to remove IPC (which removes that as the hottest codepath), I see a couple of new hot points:
![image](https://cloud.githubusercontent.com/assets/13532591/23133193/26742354-f745-11e6-8a24-b3cebf501d71.png)

These are part of https://github.com/neovim/node-client - called when it is decoded the msgpack messages. 

There are a couple of avenues to optimize this path:
- Reduce the data we send over msgpack-rpc. One place that impacts this is sending over the entire buffer on updates (`OniNotifyBufferUpdate` in `vim\core\oni-core-interop\plugin\init.vim`). Reducing this potentially decreases the amount of time we spend copying in FastBuffer, but walker is still pretty hot.
- Investigate alternatives to those libraries - for example, walker does a really  general object traversal - maybe there are some shortcuts we can take to improve the performance? If so, we can push this back to the neovim/node-client project and any other JS consumer would benefit too.

**Comment by bert88sta** [comment]
@extr0py: 

I've done a bit of preliminary work on this. I've added this to the init.vim

```
function OniNotifyBufferUpdate()

    if !exists("b:last_change_tick")
        let b:last_change_tick = -1
    endif

    if !exists("b:last_cursor_line")
        let b:last_cursor_line = 1
    endif

    if b:changedtick > b:last_change_tick
        let b:last_change_tick = b:changedtick
        let buffer_lines = getline(1,"$")
        if b:last_cursor_line < line(".")
            let buffer_lines = getline(b:last_cursor_line, line("."))
        elseif b:last_cursor_line > line(".")
           let buffer_lines = getline(line("."),b:last_cursor_line)
        "elseif b:last_cursor_line == line(".")
        "    let buffer_lines = getline(".")
        endif

        let context = OniGetContext()
        call OniNotify(["buffer_update", context, buffer_lines])
        let b:last_cursor_line = line(".")
    endif
endfunction```

Now I'm assuming that whatever handles the update logic doesn't account for it just being a string.

The commented out lines will stop oni from starting. I dont' know why or where. However, if i source the file while oni is open, I can't produce a vim error. I think the reason is because getline(".") returns a string instead of the list you expect. Unfortunately, this is the most common case for editing, where it's being typed on one line. If you could point me in the direction to fix this I think we'd see a drastic improvement

**Comment by extr0py** [comment]
Using a lighter `msgpack` library, like `msgpack-lite` might be able to help here. The `traverse`/`decode` functionality is only needed in corner cases for the returned msgpack values, but the majority of the time we don't need it, and that is pretty expensive.

Another optimization, at least on Windows, might be to communicate over sockets vs stdin/stdout.

**Comment by bert88sta** [comment]
@extr0py: Now that language server items have been implemented (mainly Change()) do you think it would be worth it to send the line/character data for buffer updates instead of the whole thing? I'm editing 700+ line scripts at work and I'm feeling the burn with the lag.

**Comment by extr0py** [comment]
Cool, that's awesome you're using Oni at work!

And yes, we should definitely look at continuing to improve performance. One key thing though is to identify the bottleneck, measure, and test whether your fix addresses that bottleneck. The chrome developer tools (included with Electron, if you use `<Control+Shift+P>`, `Open Dev Tools`, and then `Performance` tab) are really great for profiling!

Here's an example of a case that prompted me to investigate the `msgpack` layer:
![image](https://user-images.githubusercontent.com/13532591/28176391-092ffd5e-67ac-11e7-81c7-42a83d09c564.png)

I was profiling a simple scenario (open a markdown file, and use normal mode commands - `j` and `k` repeatedly). The main bottleneck was in the processing the msgpack layer does in response to the message - it could take upwards of 10ms just to process the message for a large change, and a lot of that work wasn't actually necessary.

So I think a key thing here is:
- Identify a scenario that is slow, and simplify it as much as possible. Is it that typing in insert mode is slow? Scrolling? etc
- Acquire a trace and measure the bottleneck
- Try a [tracer bullet](https://stackoverflow.com/questions/4047335/tracer-bullets-vs-prototypes) to test if the fix addresses the bottleneck

Just worth investigating, as there may be other bottlenecks depending on what your scenario is.

> I'm editing 700+ line scripts at work and I'm feeling the burn with the lag.

Ya, definitely, in many cases we send up the full file from Neovim -> Oni. There's actually work in Neovim core to enable incremental updates: https://github.com/neovim/neovim/pull/5269 - this would be awesome because it would really help that scenario! So it'd be worth leveraging that.

There are a couple of other bottlenecks I'm looking at - once the msgpack layer is addressed, it looks like the next layer is the rendering of our react components (we're re-rendering some too aggressively), and then the strategy for drawing the buffer (`DOMRenderer`) - especially in cases where a bunch of text moved, but didn't change (like scrolling or deleting a line). Past that, there's a 'built-in' latency just due to the architecture of Oni - since we have to talk to Neovim over a pipe, there's a built-in latency there. On the input side, some options would be to look at handling some keypresses (like keys in insert mode) optimistically, or even something more intense like compiling Neovim to webassembly. Fortunately (or unfortunately), there are lots of opportunities to improve performance and responsiveness here.

**Comment by bert88sta** [comment]
@extr0py : I've been back to looking at this code chunk I wrote a bit back... Who all catches the neoviminstance emit?

**Comment by extr0py** [comment]
Fixed with PR #536 - after that change, the rendering (both the editor layer and the react components) are clear bottlenecks. Onto the next round of performance optimizations 😄
